### PR TITLE
Speed-up nn.Linear for the 3d input case

### PR DESCRIPTION
**Pull request by colesbury**
This adds at::_unsafe_view and uses it in matmul. The _unsafe_view
function is identical to view except that the output is not treated
like a view by the automatic differentiation code. This avoids in-place
modifications triggering the more expensive CopySlices/AsStridedBackward
behavior.

The _unsafe_view function is only safe to use on temporaries that will
be immediately discarded and that do not alias other tensors. Otherwise,
in-place modificatiions may trigger incorrect gradients. The funciton is
not exposed to Python.

See #5169

cc @ngimel @csarofeen

### CODE REVIEW

**Review by gchanan — APPROVED** [review]
lgtm.  What do you think about adding something to the function name describing why it's unsafe?   unsafe_view_tmp_nonalias or something?  I'm sure you can come up with a better name.
